sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsGdsStore.scala

```diff
@@ -0,0 +1,128 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids
+
+import java.io.File
+import java.util.concurrent.ConcurrentHashMap
+
+import ai.rapids.cudf._
+import com.nvidia.spark.rapids.StorageTier.StorageTier
+import com.nvidia.spark.rapids.format.TableMeta
+
+import org.apache.spark.sql.rapids.RapidsDiskBlockManager
+import org.apache.spark.sql.types.DataType
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+/** A buffer store using GPUDirect Storage (GDS). */
+class RapidsGdsStore(
+    diskBlockManager: RapidsDiskBlockManager,
+    catalog: RapidsBufferCatalog = RapidsBufferCatalog.singleton)
+    extends RapidsBufferStore("gds", catalog) {
+  private[this] val sharedBufferFiles = new ConcurrentHashMap[RapidsBufferId, File]
+
+  override def createBuffer(
+      other: RapidsBuffer,
+      stream: Cuda.Stream): RapidsBufferBase = {
+    val otherBuffer = other.getMemoryBuffer
+    try {
+      val deviceBuffer = otherBuffer match {
+        case d: DeviceMemoryBuffer => d
+        case _ => throw new IllegalStateException("copying from buffer without device memory")
+      }
+      val id = other.id
+      val path = if (id.canShareDiskPaths) {
+        sharedBufferFiles.computeIfAbsent(id, _ => id.getDiskPath(diskBlockManager))
+      } else {
+        id.getDiskPath(diskBlockManager)
+      }
+      val fileOffset = if (id.canShareDiskPaths) {
+        // only one writer at a time for now when using shared files
+        path.synchronized {
+          CuFile.copyDeviceBufferToFile(path, deviceBuffer, true)
+        }
+      } else {
+        CuFile.copyDeviceBufferToFile(path, deviceBuffer, false)
+      }
+      logDebug(s"Spilled to $path $fileOffset:${other.size} via GDS")
+      new RapidsGdsBuffer(id, fileOffset, other.size, other.meta, other.getSpillPriority)
+    } finally {
+      otherBuffer.close()
+    }
+  }
+
+  class RapidsGdsBuffer(
+      id: RapidsBufferId,
+      fileOffset: Long,
+      size: Long,
+      meta: TableMeta,
+      spillPriority: Long) extends RapidsBufferBase(id, size, meta, spillPriority) {
+    private[this] var deviceBuffer: Option[DeviceMemoryBuffer] = None
+
+    override val storageTier: StorageTier = StorageTier.GDS
+
+    private def getDeviceBuffer: DeviceMemoryBuffer = synchronized {
+      if (deviceBuffer.isEmpty) {
+        val path = if (id.canShareDiskPaths) {
+          sharedBufferFiles.get(id)
+        } else {
+          id.getDiskPath(diskBlockManager)
+        }
+        val buffer = DeviceMemoryBuffer.allocate(size);
+        CuFile.copyFileToDeviceBuffer(buffer, path, fileOffset)
```

Comment:
Could `CuFile.copyFileToDeviceBuffer` fail? If so we should wrap using `closeOnExcept`:
```
closeOnExcept(DeviceMemoryBuffer.allocate(size)) { buffer => 
   CuFile.copyFileToDeviceBuffer(buffer, path, fileOffset)
}
```

Reply:
Done.